finality-aleph/src/testing/client_chain_builder.rs

```diff
@@ -0,0 +1,123 @@
+            client,
+            client_builder,
+            unique_seed: 0,
+        }
+    }
+
+    /// Import block in test client
+    pub async fn import_block(&mut self, block: Block, finalize: bool) {
+        if finalize {
+            self.client.import_as_final(BlockOrigin::Own, block.clone())
+        } else {
+            self.client.import(BlockOrigin::Own, block.clone())
+        }
+        .await
+        .unwrap();
+    }
+
+    /// Finalize block with given hash without providing justification.
+    pub fn finalize_block(&self, hash: &H256) {
+        self.client
+            .finalize_block(BlockId::Hash(*hash), None)
+            .unwrap();
+    }
+
+    pub fn genesis_hash_num(&self) -> BlockHashNum<Block> {
+        assert_eq!(
+            self.client.info().genesis_hash,
+            self.client_builder.info().genesis_hash
+        );
+        BlockHashNum::<Block>::new(self.client.info().genesis_hash, 0u64)
+    }
+
+    pub fn genesis_hash(&self) -> H256 {
+        self.genesis_hash_num().hash
+    }
+
+    pub fn get_unique_bytes(&mut self) -> Vec<u8> {
+        self.unique_seed += 1;
+        self.unique_seed.to_be_bytes().to_vec()
+    }
+
+    pub async fn build_block_at_hash(&mut self, hash: &H256) -> Block {
+        let unique_bytes: Vec<u8> = self.get_unique_bytes();
+        let mut digest = Digest::default();
+        digest.push(sp_runtime::generic::DigestItem::Other(unique_bytes));
+        let block = self
+            .client_builder
+            .new_block_at(&BlockId::Hash(*hash), digest, false)
+            .unwrap()
+            .build()
+            .unwrap()
+            .block;
+
+        self.client_builder
+            .import(BlockOrigin::Own, block.clone())
+            .await
+            .unwrap();
+        block
+    }
+
+    /// Builds a sequence of blocks extending from `hash` of length `len`
+    pub async fn build_branch_upon(&mut self, hash: &H256, len: usize) -> Vec<Block> {
+        let mut blocks = Vec::new();
+        let mut prev_hash = *hash;
+        for _i in 0..len {
+            let block = self.build_block_at_hash(&prev_hash).await;
+            prev_hash = block.hash();
+            blocks.push(block);
+        }
+        blocks
+    }
+
+    /// imports a sequence of blocks, should be in correct order
+    pub async fn import_branch(&mut self, blocks: Vec<Block>, finalize: bool) {
+        for block in blocks {
+            self.import_block(block.clone(), finalize).await;
+        }
+    }
+
+    /// Builds a sequence of blocks extending from `hash` of length `len` and imports them
+    pub async fn build_and_import_branch_upon(
+        &mut self,
+        hash: &H256,
+        len: usize,
+        finalize: bool,
+    ) -> Vec<Block> {
+        let blocks = self.build_branch_upon(hash, len).await;
+        self.import_branch(blocks.clone(), finalize).await;
+        blocks
+    }
+
+    pub fn get_header_at(&self, num: u64) -> Header {
+        self.client_builder
```

Comment:
`client` is the "main" client and the purpose of `client_builder` is creating blocks, possibly way beyond what was imported in `client`. The reason for this is technical -- you cannot create block `k+1` without importing block `k`, but we quite often want to do that.

Reply:
Added a comment

Reply:
okay, that makes sense; however, since we use the builder client in a purely 'getter' method (`get_header_at`), we must rely on the additional assumption, that the blocks known by `client_builder` form a superset of the blocks known by `client`, right? if so, then I would just add to the comment you have already introduced a short note, that _all_ the blocks have to go through `client_builder` (or more precisely, every block is created by `client_builder`)

Reply:
Added some more comments and also an assert making sure that when calling `new` the provided clients don't have any blocks imported, because it is expected here.